File: Leetcoding-Actions/Explore-Monthly-Challenges/2020-12/03-increasingOrderSearchTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode increasingBST(TreeNode root) {
        List<TreeNode> nodes = new ArrayList<>();
        Stack<TreeNode> stack = new Stack<>();
        TreeNode curr = root;
        while (true) {
            while (curr != null) {
                stack.push(curr);
                curr = curr.left;
            }
            if (stack.isEmpty()) break;
            curr = stack.pop();
            nodes.add(curr);
            curr = curr.right;
        }
        for (int i = 0; i < nodes.size() - 1; i++) {
            nodes.get(i).left = null;
            nodes.get(i).right = nodes.get(i+1);
        }
        nodes.get(nodes.size() - 1).left = null;
        nodes.get(nodes.size() - 1).right = null;
        
        return nodes.get(0);
    }
}

File: Leetcoding-Actions/Explore-Monthly-Challenges/2020-12/03-increasingOrderSearchTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def increasingBST(self, root: TreeNode) -> TreeNode:
        '''
        T: O(n) and S: O(n)
        '''
        
        # Iterative inorder
        nodes = []
        stack = []
        curr = root
        while True:
            while curr:
                stack.append(curr)
                curr = curr.left
            if not stack: break
            
            curr = stack.pop()
            nodes.append(curr)
            curr = curr.right
        
        '''
        # Recursive inorder
        nodes = []
        def inOrder(root):
            if root:
                inOrder(root.left)
                nodes.append(root)
                inOrder(root.right)
        inOrder(root)
        '''
        
        for i in range(len(nodes) - 1):
            nodes[i].right = nodes[i+1]
            nodes[i].left = None
        nodes[-1].left = None
        nodes[-1].right = None
        return nodes[0]
